examples/sonoff_basic/main.c

```diff
@@ -0,0 +1,138 @@
+ * in this order, beginning in the (square) pin header
+ * next to the button.
+ * Next hold down the button and connect the FTDI adapter
+ * to your computer. The sonoff is now in flash mode and
+ * you can flash the custom firmware.
+ *
+ * WARNING: Do not connect the sonoff to AC while it's
+ * connected to the FTDI adapter! This may fry your
+ * computer and sonoff.
+ *
+ */
+
+#include <stdio.h>
+#include <esp/uart.h>
+#include <esp8266.h>
+#include <FreeRTOS.h>
+#include <task.h>
+
+#include <homekit/homekit.h>
+#include <homekit/characteristics.h>
+#include <wifi_config.h>
+
+// The GPIO pin that is connected to the relay on the Sonoff Basic.
+const int relay_gpio = 12;
+// The GPIO pin that is connected to the LED on the Sonoff Basic.
+const int led_gpio = 13;
+// The GPIO pin that is oconnected to the button on the Sonoff Basic.
+const int button_gpio = 0;
+
+void relay_write(bool on) {
+    gpio_write(relay_gpio, on ? 1 : 0);
+}
+
+void led_write(bool on) {
+    gpio_write(led_gpio, on ? 0 : 1);
+}
+
+void switch_on_callback(homekit_characteristic_t *_ch, homekit_value_t on, void *context);
+
+void button_intr_callback(uint8_t gpio);
+
+homekit_characteristic_t switch_on = HOMEKIT_CHARACTERISTIC_(
+    ON, false, .callback=HOMEKIT_CHARACTERISTIC_CALLBACK(switch_on_callback)
+);
+
+void gpio_init() {
+    gpio_enable(led_gpio, GPIO_OUTPUT);
+    led_write(false);
+    gpio_enable(relay_gpio, GPIO_OUTPUT);
+    relay_write(switch_on.value.bool_value);
+    
+    gpio_set_pullup(button_gpio, true, true);
+    gpio_set_interrupt(button_gpio, GPIO_INTTYPE_EDGE_ANY, button_intr_callback);
+}
+
+void switch_on_callback(homekit_characteristic_t *_ch, homekit_value_t on, void *context) {
+    relay_write(switch_on.value.bool_value);
+}
+
+void button_intr_callback(uint8_t gpio) {
+    if (gpio_read(button_gpio) == 1) {
+        //Toggle the sonoff
+        switch_on.value.bool_value = !switch_on.value.bool_value;
+        relay_write(switch_on.value.bool_value);
+        homekit_characteristic_notify(&switch_on, switch_on.value);
+    }
+}
+
+void led_identify_task(void *_args) {
+    // We identify the Sonoff by Flashing it's LED.
+    for (int i=0; i<3; i++) {
+        for (int j=0; j<2; j++) {
+            led_write(true);
+            vTaskDelay(100 / portTICK_PERIOD_MS);
+            led_write(false);
+            vTaskDelay(100 / portTICK_PERIOD_MS);
+        }
+
+        vTaskDelay(250 / portTICK_PERIOD_MS);
+    }
+
+    led_write(false);
+
+    vTaskDelete(NULL);
+}
+
+void led_identify(homekit_value_t _value) {
+    printf("LED identify\n");
+    xTaskCreate(led_identify_task, "LED identify", 128, NULL, 2, NULL);
+}
+
+homekit_accessory_t *accessories[] = {
+    HOMEKIT_ACCESSORY(.id=1, .category=homekit_accessory_category_switch, .services=(homekit_service_t*[]){
+        HOMEKIT_SERVICE(ACCESSORY_INFORMATION, .characteristics=(homekit_characteristic_t*[]){
+            HOMEKIT_CHARACTERISTIC(NAME, "Sonoff"),
```

Comment:
I think this NAME is what users will see when they will start pairing with accessory. After having paired, they will have to choose individual services (like SWITCH service). So name "Sonoff" is kind of too generic to see in the list of accessories to pair. It would be better to also name it "Sonoff Switch".